secrets/certregistry/certregistry.go

```diff
@@ -0,0 +1,218 @@
+package certregistry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	defaultHost = "ingress.local"
+	currentTime = time.Now
+	// ErrNoMatchingCertificateFound is used if there is no matching certificate found
+	ErrNoMatchingCertificateFound = errors.New("no matching certificate found")
+)
+
+const (
+	// used as wildcard char in Cert Hostname/AltName matches
+	glob = "*"
+	// minimal time period for the NotAfter attribute of a Cert to be in the future
+	minimalCertValidityPeriod = 7 * 24 * time.Hour
+)
+
+type tlsCertificate struct {
+	hosts []string
+	cert  *tls.Certificate
+}
+
+// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// We ensure ensure syncronized access to them and hold a default certificate.
+type CertRegistry struct {
+	lookup         map[string]*tlsCertificate
+	mx             *sync.Mutex
+	defaultTLSCert *tls.Certificate
```

Comment:
Why do we need this fake cert?
Can we use existing `tls-cert` and `tls-key` flags for this?

Reply:
> Can we use existing
e.g. add a method to set default cert and call it from config setup

Reply:
I agree we should have a method to load a custom default certificate in the registry in case the user does not want an autogenerated one. However, requiring a TLS cert key/pair just to use Ingress TLS does not seem like a good experience. I think we should still generate a default cert in case `tls-cert` and `tis-key` are not loaded. That way users can still enable Ingress TLS but not have the key/pair prerequisite. 

This is also how `ingress-nginx` works [it generates a self-signed certificate](https://kubernetes.github.io/ingress-nginx/user-guide/tls/#default-ssl-certificate) if the `--default-ssl-certificate` secret is not specified.

Reply:
> This is also how ingress-nginx works [it generates a self-signed certificate](https://kubernetes.github.io/ingress-nginx/user-guide/tls/#default-ssl-certificate) if the --default-ssl-certificate secret is not specified.

What do you think of adding a similar option to specify a secret to load as a default tls cert?

Reply:
> However, requiring a TLS cert key/pair just to use Ingress TLS does not seem like a good experience. I think we should still generate a default cert in case tls-cert and tis-key are not loaded.

IMO certificate generation is not the main Skipper business. We do not have to introduce any magic constants like defaultHost, default validity, key type etc and there always will be users that want something slightly different.

I think it is fine and versatile to defer default certificate creation to the user as they please and have no default certificate if `tls-cert` and `tis-key` are not set.

Reply:
I agree with Alexander.
We can then also move getFakeHostTLSCert() to test if we need it for test purposes. Less code is more good.

Reply:
Agreed. I removed all the code related to default certificate generation. From the [crypto.tls config docs](https://pkg.go.dev/crypto/tls#Config) we see that: 
> GetCertificate returns a Certificate based on the given ClientHelloInfo. It will only be called if the client supplies SNI information or if Certificates is empty. If GetCertificate is nil or returns nil, then the certificate is retrieved from NameToCertificate. If NameToCertificate is nil, the best element of Certificates will be used.

So we're able to load tls certs if defined in `tls-cert` and `tis-key` as defaults if no cert is found in the registry.